### ISSUE TITLE
Error not clearing

### ISSUE THREAD

**Issue by DonRichards** [mjordan/islandora_workbench#202]
I’m seeing something interesting that might be a bug. It might also might be user error.

Create an error, fix the error and it doesn’t pickup the correction
I added `subject:Pink|subject:Flowers` to the __field_subject__ 

<img width="1568" alt="Screen Shot 2021-01-14 at 2 42 00 PM" src="https://user-images.githubusercontent.com/2738244/104641062-48f45080-5677-11eb-9e7f-618a6582e621.png">

[My Google sheet](https://docs.google.com/spreadsheets/d/1u9Jc6yC21p1je1SqiO7Fm_jliUo09K38dl7bjJAo0Bs/edit?usp=sharing)

## To replicate Error
* Add `subject:Pink|subject:Flowers` to the __field_subject__ 
* Run `./workbench --config hackdoc_demo.yml --check`
* Remove `subject:Pink|subject:Flowers` to the __field_subject__ 
* Run `./workbench --config hackdoc_demo.yml --check`

May not be related but I tried to run it without the check.

**Comment by mjordan** [comment]
Thanks @DonRichards. I'll look into this, and the first thing I'll check is whether I can replicate it using a modified local CSV instead of a Google Sheet. There might be some caching happening on the Google side.

**Comment by Digital-Grinnell** [comment]
Just wanted to add that I saw the same behavior during the hackdoc_demo, and I think Martha T. did too.  In my case I'd added a _subject:Iowa_ term and the check properly indicated that it did not exist.  When I removed that term the error was still present.  I noticed that my google_sheet.csv file did not change so the errant subject was still there.

The contents of `google_sheet.csv`, since it's a small file:

id,file,title,field_model,field_description,field_edtf_date,field_linked_agent,field_subject
101,Screen Shot 2021-01-14 at 12.22.47 PM.png,[Mark M.] Didn't bring my own image.,25,No longer a view from Deep Cove to Burnaby Mountain.  Very few mountains here in Iowa.,2021,relators:pht:person:IDK,**subject:Iowa|s**ubject:Rocks
002,IMG_5083.JPG,[Mark J] Alcatraz Island,25,"Taken from Fisherman's Wharf, San Francisco.",2014?,"relators:pht:person:Cantelo, Michele",subject:Prisons

**Comment by mjordan** [comment]
To eliminate the possibility that this ia caused by some sort of caching on the part of Google, could you try eliminating Google from the picture by replacing / commenting out with a `#` the `input_csv` entry pointing to the Sheet in your config file:

```yml
input_csv: google_sheet.csv
```
Then modify that raw CSV file and see if the modification shows up in the `--check` output? @Digital-Grinnell's note makes me thing that Google is not supplying the fresh data, or that Workbench is not retriving it over HTTP. Also, it is my understanding that Sheets "saves" data only after your cursor has left the altered cell, so if you're still in the cell, the changed data might not be ready for fetching.

**Comment by Digital-Grinnell** [comment]
Good call, Mark!  I changed my config to read `input_csv: google_sheet.csv` and removed the errant "subject:Iowa" from the data, replacing it with "subject|Mountain".  The subsequent --check did indeed change with no mention of "subject:Iowa".

So, as you suggested, I suspect this is a caching issue.

**Comment by mjordan** [comment]
Thanks for checking. I'll poke around a bit more to make sure that workbench is at least requesting the CSV from Google (should be but I'll confirm) and if we can't control the caching (maybe can via HTTP headers, etc.) at least we can document this behavior.

**Comment by mjordan** [comment]
Best predictable, replicable pattern I can see is that Sheets doesn't save changes until the cell loses focus (i.e., your cursor leaves it). Can you guys try to replicate the pattern? To test:

1. introduce a simple change that will cause workbench to detect an error, like add a single character to a vocabulary namespace
1. run workbench with `--check` to detect the error
1. fix the error, but leave your cursor in place (do not leave that cell)
1. rerun workbench, error should still be there
1. remove cursor from cell, thereby "saving" the sheet
1. rinse, repeat, observe

I was able to find one reference to this behavior, https://stackoverflow.com/questions/49657841/apps-script-force-blur-on-editing-cell-to-save-its-content, but I had hoped to find more.

**Comment by Digital-Grinnell** [comment]
Pretty sure I moved my cursor to other cells when I was doing this yesterday, but I'll give it another deliberate try now using the steps you outlined above.

**Comment by Digital-Grinnell** [comment]
Ok, results of my test...

At this moment my Google Sheet says "Last edit was yesterday at 1:35 PM".  There is NO mention of "Iowa" in my data at this time.
First test run...
╭─mark@Marks-MacBook-Pro ~/GitHub/islandora_workbench ‹main*›
╰─$ ./workbench --config hackdoc_demo2.yml --check
OK, connection to Drupal at http://97.107.189.30 verified.
OK, configuration file has all required values (did not check for optional values).
Extracting CSV data from https://docs.google.com/spreadsheets/d/12QezZCbW2ZIO8WJd8kpMgnt6OI-JVSXQ-rTs-0LA4mI/edit?usp=sharing to input_data/google_sheet.csv.
OK, CSV file input_data/google_sheet.csv found.
OK, all 2 rows in the CSV file have the same number of columns as there are headers (8).
OK, CSV column headers match Drupal field names.
OK, required Drupal fields are present in the CSV file.
OK, ETDF field values in the CSV file validate.
Error: CSV field "field_subject" in row 1 contains a term ("Iowa") that is not in the referenced vocabulary ("subject").

I made a change to a single field in the Google Sheet and the status updated to "Last edit was seconds ago", but it did so only AFTER the changed cell lost focus.  Trying another --check now.

╭─mark@Marks-MacBook-Pro ~/GitHub/islandora_workbench ‹main*›
╰─$ ./workbench --config hackdoc_demo2.yml --check                                                                                                                                              1 ↵
OK, connection to Drupal at http://97.107.189.30 verified.
OK, configuration file has all required values (did not check for optional values).
Extracting CSV data from https://docs.google.com/spreadsheets/d/12QezZCbW2ZIO8WJd8kpMgnt6OI-JVSXQ-rTs-0LA4mI/edit?usp=sharing to input_data/google_sheet.csv.
OK, CSV file input_data/google_sheet.csv found.
OK, all 2 rows in the CSV file have the same number of columns as there are headers (8).
OK, CSV column headers match Drupal field names.
OK, required Drupal fields are present in the CSV file.
OK, ETDF field values in the CSV file validate.
Error: CSV field "field_subject" in row 1 contains a term ("Iowa") that is not in the referenced vocabulary ("subject").

This is the same result as all previous attempts.  Again, my current Google Sheet makes no mention of "Iowa".

**It looks to me like `google_sheet.csv` simply isn't being rewritten as it should?**

**Comment by Digital-Grinnell** [comment]
Very strange indeed... When I look at `google_sheet.csv` I see the same data that was there yesterday, not a reflection of what's currently in my Google Sheet.  What's really strange is that prior to these tests I manually altered `google_sheet.csv` and removed references to "Iowa", but they are back.

I'm going to delete change `google_sheet.csv` again, save the change and run the --check again to see what happens.

**Comment by DonRichards** [comment]
@mjordan This light help, looks like it's related and with a solution https://stackoverflow.com/a/17347290

**Comment by Digital-Grinnell** [comment]
OK, I think I see the source of my woes...

My config file looks like this:

```
task: create
host: "http://97.107.189.30"
username: admin
password: islandora
Allow_adding_terms: true
Google_sheets_gid: 1094504353
# input_csv: google_sheet.csv
input_csv: 'https://docs.google.com/spreadsheets/d/12QezZCbW2ZIO8WJd8kpMgnt6OI-JVSXQ-rTs-0LA4mI/edit?usp=sharing'
```

**However, this process appears to be loading data from my "Task 3" tab, not "Task 4", where I have been editing.  My understanding was that the `Google_sheets_gid: 1094504353` would direct the input to come from "Task 4", but perhaps I have done something incorrectly here?**

Would it be possible to put the name of the tab in such a field, rather than the `gid`?  That's what we do in IMI by the way.

**Comment by Digital-Grinnell** [comment]
Just now read the link that @mjordan provided.  That does look troublesome.  Mark, in case you plan to make a change to your parameters, as suggested in the linked post, I'll keep my test case at-the-ready.

**Comment by Digital-Grinnell** [comment]
Another clue...  I put the text "subject:Task3" and "subject:Task4" into appropriate cells in my "Task 3" and "Task 4" tabs, then ran another --check:

```
╭─mark@Marks-MacBook-Pro ~/GitHub/islandora_workbench ‹main*›
╰─$ ./workbench --config hackdoc_demo2.yml --check                                                                                                                                              1 ↵
OK, connection to Drupal at http://97.107.189.30 verified.
OK, configuration file has all required values (did not check for optional values).
Extracting CSV data from https://docs.google.com/spreadsheets/d/12QezZCbW2ZIO8WJd8kpMgnt6OI-JVSXQ-rTs-0LA4mI/edit?usp=sharing to input_data/google_sheet.csv.
OK, CSV file input_data/google_sheet.csv found.
OK, all 2 rows in the CSV file have the same number of columns as there are headers (8).
OK, CSV column headers match Drupal field names.
OK, required Drupal fields are present in the CSV file.
OK, ETDF field values in the CSV file validate.
Error: CSV field "field_subject" in row 1 contains a term ("Task3") that is not in the referenced vocabulary ("subject").
```

Clearly my `Google_sheets_gid: 1094504353` parameter is not directing input from "Task 4" as intended.

**Comment by mjordan** [comment]
I opened #206 this morning, gonna have to investigate that one. 

@DonRichards thanks for the link, but Workbench is not using the Python API to Google Sheets, it's just dumpting the CSV directly via a special URL. I investigated the API and it requires OAuth, etc. which will introduce another set of headaches, but if we need to use it, we can. I'm going to resolve #206 first before considering anything else though.

**Comment by mjordan** [comment]
And ^^  https://stackoverflow.com/a/17347290 uses the Javascript API. Will continue to investigate, please keep examples/hunches coming as your time allows.

**Comment by mjordan** [comment]
@DonRichards, in #206 @Digital-Grinnell  pointed out that an uppercase letter in the `google_sheets_gid` config key was likely his issue, and Workbench now converts all config keys to lower case. Can you check to see if that solves your experiences described above?

**Comment by DonRichards** [comment]
@mjordan I'll give it a go.

**Comment by DonRichards** [comment]
@mjordan that did not correct the issue until I pulled the updates. Appears to be fine now. I do see a warning now. 

> Warning: Issues detected with validating typed relation field values in the CSV file. See the log for more detail.

**Comment by Digital-Grinnell** [comment]
@mjordan I also gave this a test and confirm the same.  Before pulling the new code my `Google_sheet_gid` parameter appears to have been ignored.  After pulling the new code the same parameter is clearly working as intended.  Thanks for the improvements!

**Comment by mjordan** [comment]
@DonRichards @Digital-Grinnell thanks very much for testing. I'm going to close this now since it seems to be resolved but if anything else comes up feel free to open  new issue.
